Improve setup

kenjis opened the pull request:
Follow-up #368

- show Skip message once
- small fix for the docs

---

Clean installation and
When `$this->helpers = array_merge($this->helpers, ['auth', 'setting']);` exists: 
![screenshot 2022-08-14 19 32 06](https://user-images.githubusercontent.com/87955/184532916-a1b2171a-6cd8-4bb3-a2ee-434330d3dcc1.png)

---

When `$this->helpers = array_merge($this->helpers, ['setting']);` exists:
![screenshot 2022-08-14 19 33 36](https://user-images.githubusercontent.com/87955/184532966-fabfb591-787f-4282-ae64-8d26e27446a3.png)

datamweb reviewed (approved):
Great, I confirm that the problem is fixed.
![Screenshot 2022-08-14 142303](https://user-images.githubusercontent.com/9530214/184533548-9a635c4d-af5f-4d37-9a5d-10d35d9c843f.jpg)
Thanks!

MGatner commented on src/Commands/Setup.php:
```diff
@@ -230,7 +234,10 @@ private function setupHelper(): void
             '$this->helpers = array_merge($this->helpers, [\'auth\', \'setting\']);' => $check,
         ];
-        $this->replace($file, $replaces);
+        $return = $this->replace($file, $replaces);
+        if ($return) {
```
Since it is always a boolean can we skip the intermediary?
```suggestion
        if ($this->replace($file, $replaces)) {
```

datamweb commented:
Wait please!

kenjis commented:
@datamweb What do you mean? Any problem?

datamweb commented:
> @datamweb What do you mean? Any problem?

@kenjis, `APPPATH\Config/Auth.php` Or `APPPATH\Config\Auth.php`?

kenjis commented:
No problem. The code works. It is just only the output is a bit strange.
It is because Windows uses `\` as a path separator.

On Unix system:
```
CodeIgniter v4.2.4 Command Line Tool - Server Time: 2022-08-15 02:57:04 UTC-05:00

  Created: APPPATH/Config/Auth.php
  Created: APPPATH/Config/AuthGroups.php
  Updated: APPPATH/Controllers/BaseController.php
  Updated: APPPATH/Config/Routes.php
  Run `spark migrate --all` now? [y, n]: n
```

datamweb commented:
Okey @Kenjis, my question is why when there are `app\Config\Auth.php` and `app\Config\AuthGroups.php` files.
Is the correct output shown?
```
  File 'APPPATH\Config\Auth.php' already exists in destination. Overwrite? [n, y]: y
  Created: APPPATH\Config\Auth.php
  File 'APPPATH\Config\AuthGroups.php' already exists in destination. Overwrite? [n, y]: y
  Created: APPPATH\Config\AuthGroups.php
  Skipped APPPATH\Controllers\BaseController.php. It has already been updated.
  Skipped APPPATH\Config\Routes.php. It has already been updated.
  Run `spark migrate --all` now? [y, n]: n
```

datamweb commented:
@kenjis , I found a way, but I don't know if it's the right one.
If you change `/` to` \` below, the correct output will be shown.
https://github.com/codeigniter4/shield/blob/c48cb81bd80762e0133d1ba53f50e16ab99c1f6e/src/Commands/Setup.php#L108

https://github.com/codeigniter4/shield/blob/c48cb81bd80762e0133d1ba53f50e16ab99c1f6e/src/Commands/Setup.php#L120

output:
```
P:\WorkApp\CI423\cI423>php spark shield:setup

CodeIgniter v4.2.4 Command Line Tool - Server Time: 2022-08-15 04:03:36 UTC-05:00

  Created: APPPATH\Config\Auth.php
  Created: APPPATH\Config\AuthGroups.php
  Skipped APPPATH\Controllers\BaseController.php. It has already been updated.
  Skipped APPPATH\Config\Routes.php. It has already been updated.
  Run `spark migrate --all` now? [y, n]:
```

kenjis commented:
>  I found a way, but I don't know if it's the right one.

It works only on Windows.

kenjis commented:
> why when there are app\Config\Auth.php and app\Config\AuthGroups.php files.
Is the correct output shown?

I don't know. It is a mystery!
Search "Created:" in the `Setup` class. There is only one place.

MGatner reviewed (approved):
Looks good to me! I don't really care about the path separator issue as I think the message conveys its intent both ways, but I'm happy for you two to keep discussing 🤗

kenjis commented:
@datamweb The solution that works perfectly is to use `DIRECTORY_SEPARATOR` for the path in PHP code.
```php
$file     = 'Config' . DIRECTORY_SEPARATOR . 'Auth.php'
```
But I don't like it, because the code will be a bit difficult to read and longer.

datamweb commented:
@kenjis, Even if we define a private function, for example
```php
$file     = 'Config/Auth.php';
$file = $this->modifyDisplayPath($file);
```
Do you believe the code is unreadable?
If yes, I have no problem with the merge.

kenjis commented:
@datamweb Yes, it is a way to handle this. Where should we do the convertion?

datamweb commented:
@kenjis, that's what I meant.

```php
    private function publishConfigAuth(): void
    {
        
         $file = 'Config/Auth.php';
         $file = $this->modifyDisplayPath($file);
        //$file     = 'Config' . DIRECTORY_SEPARATOR . 'Auth.php';
        $replaces = [
            'namespace CodeIgniter\Shield\Config'  => 'namespace Config',
            'use CodeIgniter\\Config\\BaseConfig;' => 'use CodeIgniter\\Shield\\Config\\Auth as ShieldAuth;',
            'extends BaseConfig'                   => 'extends ShieldAuth',
        ];

        $this->copyAndReplace($file, $replaces);
    }
    
    private function modifyDisplayPath(string $file): string
    {
        return str_replace('/',DIRECTORY_SEPARATOR,$file);
    }
```

kenjis commented:
There is another way, to convert just right before output.

datamweb commented:
```diff
    protected function writeFile(string $file, string $content): void
    {
        $path      = $this->distPath . $file;
--- $cleanPath = clean_path($path);
+++  $cleanPath = $this->modifyDisplayPath(clean_path($path));

        $directory = dirname($path);
```
```diff
+++private function modifyDisplayPath(string $file): string
+++{
+++return str_replace('/',DIRECTORY_SEPARATOR,$file);
+++}
```

MGatner commented:
Would it be too heavy-handed to modify `clean_path()` to do this natively? Convert between OS-specific slashes.

kenjis commented:
I think that is possible.
